File: 2024/02/main.py
#!/usr/bin/env python3

import sys

def doPart1():
    filename = "example.dat"
    if len(sys.argv) > 1:
        filename = sys.argv[1]

    runningSum = 0
    with open(filename, "r") as f:
        for li in f:
            line = li.strip()
            if not line:
                continue
            sLine = line.split()
            lastVal = -1
            isSafe = 0
            isPos = 0
            isNeg = 0
            canRemove = 0
            for cStr in sLine:
                val = int(cStr)
                if lastVal < 0:
                    lastVal = val
                    continue
                if val - lastVal > 0:
                    isPos = 1
                else:
                    isNeg = 1
                if isPos and isNeg:
                    if canRemove:
                        canRemove = 0
                    else:
                        break
                if (isPos):
                    a = lastVal
                    b = val
                else:
                    a = val
                    b = lastVal
                lastVal = val
                if (b-a > 3 or b==a):
                    if canRemove:
                        canRemove = 0
                    else:
                        break
            else:
                isSafe = 1
            runningSum += isSafe
        print(runningSum)

def checkSafe(sLine:list[str]) -> int:
    lastVal = -1
    isPos = 0
    isNeg = 0
    for cStr in sLine:
        val = int(cStr)
        if lastVal < 0:
            lastVal = val
            continue
        if val - lastVal > 0:
            isPos = 1
        elif val - lastVal < 0:
            isNeg = 1
        else:
            return 0
        if isPos and isNeg:
            return 0
        if (isPos):
            a = lastVal
            b = val
        else:
            a = val
            b = lastVal
        if (b-a > 3 or b==a):
            return 0
        lastVal = val

    return 1


def isSafeFun(sLine:list[str]):
    # Check if the original list is safe
    isSafe = checkSafe(sLine)
    if isSafe:
        return 1
    for k in range(len(sLine)):
        nLine = sLine[:k] + sLine[(k+1):]
        isSafe = checkSafe(nLine)
        if isSafe:
            return 1
    return 0


def doPart2():
    filename = "example.dat"
    if len(sys.argv) > 1:
        filename = sys.argv[1]

    runningSum = 0
    with open(filename, "r") as f:
        for li in f:
            line = li.strip()
            if not line:
                continue
            sLine = line.split()
            isSafe = isSafeFun(sLine)
            runningSum += isSafe
        print(runningSum)

def main():
    doPart1()
    doPart2()


if __name__ == "__main__":
    main()

File: 2024/02/main.cpp

#include "../helpers/extras.hh"

int checkSafe(std::vector<std::string>& sLine)
{
    int lastNum = -1;
    bool isIncreasing = 0;
    bool isDecreasing = 0;
    int a;
    int b;
    for (std::string c : sLine) {
        const int num = atoi(c.c_str());
        if (lastNum < 0) {
            lastNum = num;
            continue;
        }
        if (num - lastNum > 0)
            isIncreasing = 1;
        else if (num - lastNum < 0)
            isDecreasing = 1;
        else
            return 0;

        if (isIncreasing > 0 && isDecreasing > 0) {
            return 0;
        }
        if (isIncreasing) {
            a = lastNum;
            b = num;
        } else {
            b = lastNum;
            a = num;
        }
        lastNum = num;
        if (b - a > 3 || b == a) {
            return 0;
        }
    }
    return 1;
}

int isSafeFun(std::vector<std::string>& sLine)
{
    int isSafe = checkSafe(sLine);
    if (isSafe == 1) return 1;

    for (size_t k = 0; k < sLine.size(); k++) {
        std::vector<std::string> nLine;
        for (size_t k1 = 0; k1 < k; k1++) {
            nLine.push_back(sLine[k1]);
        }
        for (size_t k1 = k+1; k1 < sLine.size(); k1++) {
            nLine.push_back(sLine[k1]);
        }
        isSafe = checkSafe(nLine);
        if (isSafe == 1) return 1;
    }
    return 0;
}

void doPart1(const char* filename)
{
    std::ifstream file(filename);
    std::string line;
    int sum = 0;
    while (getline(file, line)) {
        stripString(line);
        if (line.length() == 0) continue;
        // printf("%s\n", line.c_str());
        std::vector<std::string> sLine = splitString(line, " ");
        int isSafe = checkSafe(sLine);
        sum += isSafe;
    }
    file.close();
    printf("%d\n", sum);
}

void doPart2(const char* filename)
{
    std::ifstream file(filename);
    std::string line;
    int sum = 0;
    while (getline(file, line)) {
        stripString(line);
        if (line.length() == 0) continue;
        std::vector<std::string> sLine = splitString(line, " ");
        sum += isSafeFun(sLine);
    }
    file.close();
    printf("%d\n", sum);
}

int main(int narg, char* args[])
{
    std::string filename = "example.dat";
    if (narg > 1) filename = std::string(args[1]);
    doPart1(filename.c_str());
    doPart2(filename.c_str());
    return 0;
}
